okhttp/src/main/java/okhttp3/Request.java

```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
       return this;
     }
 
+    /** Attaches {@code tag} to the request using {@code Object.class} as a key. */
+    public Builder tag(@Nullable Object tag) {
+      return tag(Object.class, tag);
+    }
+
     /**
-     * Attaches {@code tag} to the request. It can be used later to cancel the request. If the tag
-     * is unspecified or null, the request is canceled by using the request itself as the tag.
+     * Attaches {@code tag} to the request using {@code type} as a key. Tags can be read from a
+     * request using {@link Request#tag}. Use null to remove any existing tag assigned for {@code
+     * type}.
+     *
+     * <p>Use this API to attach timing, debugging, or other application data to a request so that
+     * you may read it in interceptors, event listeners, or callbacks.
      */
-    public Builder tag(Object tag) {
-      this.tag = tag;
+    public <T> Builder tag(Class<? extends T> type, @Nullable T tag) {
+      if (type == null) throw new IllegalArgumentException("type == null: " + type);
+
+      if (tag == null) {
+        tags.remove(type);
```

Comment:
Is this safe on empty map?

Reply:
No, `UnsupportedOperationException` is thrown for `Collections.emptyMap()`.
This method returns unmodifiable empty map.
This operation is safe only for empty `LinkedHashMap`

Reply:
It’s safe. I added this test.
```
  @Test public void removeAbsentTag() {
    Request request = new Request.Builder()
        .url("https://square.com")
        .tag(null)
        .build();
    assertNull(request.tag());
  }
```